### DISCUSSION TITLE
How to trigger a function every time a file is uploaded into storage?

### DISCUSSION THREAD

**Discussion by hallee9000** [supabase/supabase#6540] [Questions]
Hi everyone, I'm not skilled in Postgres. I want to implement a storage app. Users can upload files in the app. I want to charge them by how much they used.

Let's say I have a table to record everyone's left volume.

<img width="535" alt="image" src="https://user-images.githubusercontent.com/12868055/164982098-cf9c4f01-1209-49f8-adc6-1d8aa630df82.png">

I tried to write some SQL to create a trigger and function like this:

```sql
-- Reduce the left storage volume each time user upload a file
create function public.handle_personal_storage_update() 
returns trigger 
language plpgsql 
security definer set search_path = public
as $$
begin
  update public.profiles
  set storage = storage-new.metadata.size/1024
  where id = new.owner and new.bucket_id="transferred-photos";
end;
$$;

-- trigger the function every time a user uploads a file
create trigger on_auth_user_created
  after insert on storage.objects
  for each row execute procedure public.handle_personal_storage_update();
```

Now I get this error when I upload a file. Is something wrong with the SQL? Thanks in advance.

<img width="649" alt="image" src="https://user-images.githubusercontent.com/12868055/164982246-13b88e9f-0975-4f90-a974-a6db4df3de8f.png">

**Comment by GaryAustin1** [comment]
Several issues at least:
1) I would not use on_auth_user_created for the trigger name, that is the guide name for a trigger on auth.user insert.  If you happen to have that also you will run into issues, it is also not descriptive of what you are doing on the trigger.

2) "transferred-photos" should be 'transferred photos' as strings are single quote only in postgres.

3) The metadata column is jsonb.  To reference it you would do something like new.metadata->>'size'  

4) And this is the big one.  
When the file object is uploaded there are two operations to the object table.  
The first creates the object with owner and name info to verify RLS using INSERT (there is no metadata at this point).  
Then the storage API stores the file to S3 storage and gets metadata back including size.  
The storage API then does an UPDATE operation on objects table to update the metadata.  
So you will likely need to trigger on UPDATE instead AND then make sure metadata is present as I'm not sure when else storage might update the object.

**Reply by hallee9000** [reply]
Thanks, @GaryAustin1 , I just forgot to change the trigger name. Now I changed the SQL as below, but it still doesn't work.
```sql
-- Reduce the left storage volume each time user upload a file
create function public.handle_personal_storage_update() 
returns trigger 
language plpgsql 
security definer set search_path = public
as $$
begin
  update public.profiles
  set storage = storage-(new.metadata->>'size')
  where id = new.owner and new.bucket_id='transferred-photos';
end;
$$;

-- trigger the function every time a user uploads a file
create trigger on_file_updated
  after update on storage.objects
  for each row execute procedure public.handle_personal_storage_update();
```

<img width="619" alt="image" src="https://user-images.githubusercontent.com/12868055/165002280-2c5e9a58-83d9-4a78-8091-e5dd9aec57cb.png">

Why the size is a text?

**Reply by GaryAustin1** [reply]
Hmmm either something like new.metadata->>'size'::bigint or maybe new.metadata->'size'     with single arrow.  I don't use postgres jsonb enough to remember....

**Reply by hallee9000** [reply]
Great thanks @GaryAustin1 .
Finally I figured it out as below:

```sql
-- Reduce the left storage volume each time user upload a file
create function public.handle_personal_storage_update() 
returns trigger 
language plpgsql 
security definer set search_path = public
as $$
begin
  update public.profiles
  set storage = storage-(new.metadata->'size')::int
  where id = new.owner and new.bucket_id='transferred-photos';
  return new;
end;
$$;

-- trigger the function every time a user uploads a file
create trigger on_file_updated
  after update on storage.objects
  for each row execute procedure public.handle_personal_storage_update();
```

**Reply by GaryAustin1** [reply]
This is good for getting going, but you will also need to understand what other times update can occur.  For instance .from.update if you support that.  
And also deal with delete if you support that, and add back space.
Glad you are able to move ahead.

**Reply by Monu18765** [reply]
**New Update**
1. update action is not triggering when new file is inserted.
2. insert action is triggering on Deleting of file.

**Summary**
I don't think it is safe to use storage.objects table events. It can change anytime. We probably need a proper solution from Supabase to work confidently.

**Reply by GaryAustin1** [reply]
Yes in other posts over the past 6 months I've warned the CURRENT (at that time) storage works like this, and did not know if the new version would work the same way.  Always risky counting on Supabase schemas if they have not themselves documented the approach to take.  I've never seen an official method presented by them for doing this, triggers or not.

**Reply by Monu18765** [reply]
We seriously need a solution, a table maintained by supabase which we can use it as source of truth.

**Reply by GaryAustin1** [reply]
There already may be such a request in the supabase/storage-api github, but that would be the place to generate the issue.

**Reply by Monu18765** [reply]
Yeah I know but I need it right now. I am using triggers to compress images at the backend. 
I will have to work with some make shift solution.

**Reply by GaryAustin1** [reply]
Note I'm not saying it works differently now.  I've not checked.  I'm relying on your statement it is not.
Certainly there is still an insert to storage.objects for an upload and I can't imagine why they would be doing an insert to storage.
objects on delete.

Well, if I needed a solution right now, and I'm on the new storage-api...

I would either go thru the source code and see what they do on storage.objects now or hook up a trigger function on all 3 insert/update/delete.   Then use `raise log 'new = %, tg_op = %',new,TG_OP;` and look in the Postgres log for the pattern.

Unless they already have something in the works, it is not likely to happen soon, especially if there is not already a request for it in storage-api or here as an enhancement (and there might be as it is a common issue).

**Reply by Monu18765** [reply]
Yes, I did the same I have setup triggers for "after insert" and it get triggered when photo is deleted with below **new** value.
```
(704ba4a3-5c72-45f4-bfc4-2b99b0d69858,profiles,89bb2bfd-2ba7-4a3c-af84-4e98b659ccd0/.emptyFolderPlaceholder,,"2023-11-15 17:27:18.715577+00","2023-11-15 17:27:18.715577+00","2023-11-15 17:27:18.715577+00",,"{89bb2bfd-2ba7-4a3c-af84-4e98b659ccd0,.emptyFolderPlaceholder}",1,)
```

I will suggest you give it a try.

**Reply by GaryAustin1** [reply]
I don't use triggers in storage objects for my code, so probably will not follow up on this for awhile.
Unless the new storage is default, I likely don't even have it on any of my instances.

**Reply by GaryAustin1** [reply]
It is best to raise your concerns with Supabase directly in storage-api.

**Reply by Monu18765** [reply]
Got it.

**Comment by GaryAustin1** [comment]
The new version of storage has radically changed their trigger methods.
https://github.com/orgs/supabase/discussions/19017#discussioncomment-7613197
Note that all use of triggers on storage schema is at risk of changes...

**Reply by Monu18765** [reply]
Yes, it is a tricky path to walk and I hope we will see something solid from Supabase.

**Reply by MentalGear** [reply]
Could you share your XP with the storage triggers please, is it reliable now to use?

**Reply by Monu18765** [reply]
Hey @MentalGear so far  I  have no issue. Make sure you test for every possible event.

**Comment by logemann** [comment]
Thanks @GaryAustin1 for the heads up. I am also really wondering how people keep a metadata table with a 1:1 relation to the storage items in sync. First i thought i will do it client side with a "upload-success" event, but then i have another channel (via email) where people can send files into storage. So basically a trigger is IMO the only place where you can be sure that the code for inserting/updating metadata always works. 

And you can also delete files in the supabase console, so i also need a method to delete my related record in the metadata table in this case. To me, triggers are the only sane way to do it....

**Reply by MentalGear** [reply]
Hey @logemann, in the same spot as you were a month ago. Would you mind sharing your experience and whether you found a working solution for storage triggers?

**Comment by varun508** [comment]
@MentalGear @Monu18765 
I'm also a bit confused with this. I was thinking of running my background task at an endpoint on INSERT only when I get a metadata field.
As this [discussion](https://github.com/orgs/supabase/discussions/19017#discussioncomment-7613197), points out that it triggers with metadata

**Reply by Monu18765** [reply]
Check the new, old value for each trigger. And somehow make it work.

<img width="1074" alt="Screenshot 2024-03-19 at 1 14 04 AM" src="https://github.com/supabase/supabase/assets/144263406/6db85cd1-497e-43f9-bad7-1a55b01fd4b5">

**Reply by Monu18765** [reply]
pay attention to how eTag changes.
